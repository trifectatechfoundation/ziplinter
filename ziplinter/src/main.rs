use rc_zip_sync::ReadZip;
use std::ops::Deref;

fn main() {
    let mut args = std::env::args();
    let _ = args.next();
    let path = args.next().unwrap();

    let file = std::fs::File::open(path).unwrap();
    let archive = file.read_zip().unwrap();

    eprintln!(
        "{}",
        serde_json::to_string_pretty(&archive.deref()).unwrap()
    )
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;
    use std::{error::Error, path::Path};

    fn process_zip_file(zip_path: &Path) -> Result<serde_json::Value, Box<dyn Error>> {
        let file = std::fs::File::open(zip_path).unwrap();
        let archive = file.read_zip();

        Ok(serde_json::to_value(archive?.deref())?)
    }

    #[test]
    fn snapshot_zip_files() {
        let fixtures_dir = std::env::current_dir()
            .unwrap()
            .join("../testdata")
            .canonicalize()
            .unwrap();

        println!("fixtures_dir: {}", fixtures_dir.display());

        for entry in std::fs::read_dir(fixtures_dir).expect("Failed to read fixtures directory") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();

            if path
                .extension()
                .is_some_and(|e| e.eq_ignore_ascii_case("zip"))
            {
                let Ok(result) = process_zip_file(&path) else {
                    continue;
                };
                println!(
                    "current file: {}",
                    Path::new(path.file_name().unwrap()).display()
                );
                assert_json_snapshot!(
                    format!("{}", Path::new(path.file_name().unwrap()).display()),
                    result
                );
            }
        }
    }
}
