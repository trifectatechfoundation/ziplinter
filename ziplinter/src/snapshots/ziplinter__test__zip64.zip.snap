---
source: ziplinter/src/main.rs
expression: result
---
{
  "comment": "",
  "contents": [
    {
      "central": {
        "comment": "",
        "compressed_size": 4294967295,
        "crc32": 1778378622,
        "creator_version": {
          "host_system": "Unix",
          "version": 45
        },
        "disk_nbr_start": 0,
        "external_attrs": 2175008768,
        "extra": [
          1,
          0,
          16,
          0,
          36,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          36,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 0,
        "internal_attrs": 0,
        "method": "Deflate",
        "mode": 420,
        "modified": "2012-08-10T14:33:32Z",
        "name": "README",
        "reader_version": {
          "host_system": "MsDos",
          "version": 45
        },
        "uncompressed_size": 4294967295
      },
      "local": {
        "accessed": null,
        "compressed_size": 36,
        "crc32": 1778378622,
        "created": null,
        "extra": [],
        "flags": 0,
        "gid": null,
        "header_offset": 0,
        "method": "Deflate",
        "method_specific": "None",
        "mode": 0,
        "modified": "2012-08-10T14:33:32Z",
        "name": "README",
        "reader_version": {
          "host_system": "MsDos",
          "version": 20
        },
        "uid": null,
        "uncompressed_size": 36
      }
    }
  ],
  "encoding": "Utf8",
  "eocd": {
    "dir": {
      "inner": {
        "comment": [],
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 65535,
        "directory_offset": 4294967295,
        "directory_records": 65535,
        "directory_size": 4294967295,
        "disk_nbr": 0
      },
      "offset": 220
    },
    "dir64": {
      "inner": {
        "creator_version": 45,
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 1,
        "directory_offset": 72,
        "directory_records": 1,
        "directory_size": 72,
        "disk_nbr": 0,
        "reader_version": 45,
        "record_size": 44
      },
      "offset": 144
    },
    "global_offset": 0
  },
  "parsed_ranges": [
    {
      "contains": "end of central directory record",
      "end": 242,
      "start": 220
    },
    {
      "contains": "zip64 end of central directory locator",
      "end": 220,
      "start": 200
    },
    {
      "contains": "zip64 end of central directory record",
      "end": 200,
      "start": 144
    },
    {
      "contains": "central directory header",
      "end": 144,
      "filename": "README",
      "start": 72
    },
    {
      "contains": "local file header",
      "end": 36,
      "filename": "README",
      "start": 0
    },
    {
      "contains": "file data",
      "end": 72,
      "filename": "README",
      "start": 36
    }
  ],
  "size": 242
}
